proxymode/proxymode.go

```diff
@@ -0,0 +1,150 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package proxymode
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	"github.com/elastic/package-registry/packages"
+	"github.com/elastic/package-registry/util"
+)
+
+type ProxyMode struct {
+	options ProxyOptions
+
+	httpClient     *http.Client
+	destinationURL *url.URL
+	resolver       *proxyResolver
+}
+
+type ProxyOptions struct {
+	Enabled bool
+	ProxyTo string
+}
+
+func NoProxy() *ProxyMode {
+	proxyMode, err := NewProxyMode(ProxyOptions{})
+	if err != nil {
+		log.Fatalf("no proxy mode should not return an error: %v", err)
+	}
+	return proxyMode
+}
+
+func NewProxyMode(options ProxyOptions) (*ProxyMode, error) {
+	var pm ProxyMode
+	pm.options = options
+
+	if !options.Enabled {
+		return &pm, nil
+	}
+
+	pm.httpClient = &http.Client{
+		Timeout: 10 * time.Second,
+		Transport: &http.Transport{
+			MaxIdleConns:        100,
+			MaxIdleConnsPerHost: 100,
+			IdleConnTimeout:     90 * time.Second,
+		},
+	}
+
+	var err error
+	pm.destinationURL, err = url.Parse(pm.options.ProxyTo)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't create proxy destination URL")
+	}
+
+	pm.resolver = &proxyResolver{destinationURL: *pm.destinationURL}
+	return &pm, nil
+}
+
+func (pm *ProxyMode) Enabled() bool {
+	return pm.options.Enabled
+}
+
+func (pm *ProxyMode) Search(r *http.Request) (packages.Packages, error) {
+	if !pm.options.Enabled {
+		return packages.Packages{}, nil
+	}
+	logger := util.Logger()
+
+	proxyURL := *r.URL
+	proxyURL.Host = pm.destinationURL.Host
+	proxyURL.Scheme = pm.destinationURL.Scheme
+	proxyURL.User = pm.destinationURL.User
+
+	proxyRequest, err := http.NewRequest("GET", proxyURL.String(), nil)
```

Comment:
Nit. Use the same method in the proxy call?
```suggestion
	proxyRequest, err := http.NewRequest(r.Method, proxyURL.String(), nil)
```

Reply:
Not sure about this. I don't know if we want to pass the HEAD or POST requests.

Reply:
Good idea, fixed.